File: backend/src/main/java/com/relief/service/training/KnowledgeBaseService.java
package com.relief.service.training;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.*;

/**
 * Knowledge base service for searchable repository of best practices and procedures
 */
@Service
@RequiredArgsConstructor
@Slf4j
public class KnowledgeBaseService {

    private static final Logger log = LoggerFactory.getLogger(KnowledgeBaseService.class);

    public KnowledgeArticle createArticle(String title, String content, String category, String tags, 
                                        String authorId, String language, boolean isPublic) {
        KnowledgeArticle article = new KnowledgeArticle();
        article.setId(UUID.randomUUID().toString());
        article.setTitle(title);
        article.setContent(content);
        article.setCategory(category);
        article.setTags(Arrays.asList(tags.split(",")));
        article.setAuthorId(authorId);
        article.setLanguage(language);
        article.setIsPublic(isPublic);
        article.setCreatedAt(LocalDateTime.now());
        article.setUpdatedAt(LocalDateTime.now());
        article.setViewCount(0);
        article.setRating(0.0);
        article.setStatus(ArticleStatus.PUBLISHED);
        
        log.info("Created knowledge article: {}", article.getId());
        return article;
    }

    public KnowledgeArticle updateArticle(String articleId, String title, String content, String category, 
                                        String tags, String language, boolean isPublic) {
        KnowledgeArticle article = new KnowledgeArticle();
        article.setId(articleId);
        article.setTitle(title);
        article.setContent(content);
        article.setCategory(category);
        article.setTags(Arrays.asList(tags.split(",")));
        article.setLanguage(language);
        article.setIsPublic(isPublic);
        article.setUpdatedAt(LocalDateTime.now());
        
        log.info("Updated knowledge article: {}", articleId);
        return article;
    }

    public KnowledgeArticle getArticle(String articleId) {
        // Implementation for getting article
        KnowledgeArticle article = new KnowledgeArticle();
        article.setId(articleId);
        article.setTitle("Sample Article");
        article.setContent("Sample content");
        article.setCategory("Emergency Response");
        article.setTags(Arrays.asList("disaster", "response", "emergency"));
        article.setAuthorId("author-123");
        article.setLanguage("en");
        article.setIsPublic(true);
        article.setCreatedAt(LocalDateTime.now());
        article.setUpdatedAt(LocalDateTime.now());
        article.setViewCount(0);
        article.setRating(0.0);
        article.setStatus(ArticleStatus.PUBLISHED);
        
        return article;
    }

    public List<KnowledgeArticle> searchArticles(String query, String category, List<String> tags, 
                                               String language, boolean isPublic, int limit) {
        // Implementation for searching articles
        return Collections.emptyList();
    }

    public List<KnowledgeArticle> getArticlesByCategory(String category, int limit) {
        // Implementation for getting articles by category
        return Collections.emptyList();
    }

    public List<KnowledgeArticle> getPopularArticles(int limit) {
        // Implementation for getting popular articles
        return Collections.emptyList();
    }

    public List<KnowledgeArticle> getRecentArticles(int limit) {
        // Implementation for getting recent articles
        return Collections.emptyList();
    }

    public KnowledgeArticle rateArticle(String articleId, String userId, double rating, String comment) {
        ArticleRating articleRating = new ArticleRating();
        articleRating.setId(UUID.randomUUID().toString());
        articleRating.setArticleId(articleId);
        articleRating.setUserId(userId);
        articleRating.setRating(rating);
        articleRating.setComment(comment);
        articleRating.setCreatedAt(LocalDateTime.now());
        
        log.info("Rated article {} with rating {}", articleId, rating);
        return getArticle(articleId);
    }

    public KnowledgeArticle bookmarkArticle(String articleId, String userId) {
        ArticleBookmark bookmark = new ArticleBookmark();
        bookmark.setId(UUID.randomUUID().toString());
        bookmark.setArticleId(articleId);
        bookmark.setUserId(userId);
        bookmark.setCreatedAt(LocalDateTime.now());
        
        log.info("Bookmarked article {} by user {}", articleId, userId);
        return getArticle(articleId);
    }

    public void removeBookmark(String articleId, String userId) {
        log.info("Removed bookmark for article {} by user {}", articleId, userId);
    }

    public List<KnowledgeArticle> getUserBookmarks(String userId) {
        // Implementation for getting user bookmarks
        return Collections.emptyList();
    }

    public List<KnowledgeArticle> getRelatedArticles(String articleId, int limit) {
        // Implementation for getting related articles
        return Collections.emptyList();
    }

    public KnowledgeCategory createCategory(String name, String description, String parentCategoryId) {
        KnowledgeCategory category = new KnowledgeCategory();
        category.setId(UUID.randomUUID().toString());
        category.setName(name);
        category.setDescription(description);
        category.setParentCategoryId(parentCategoryId);
        category.setCreatedAt(LocalDateTime.now());
        category.setIsActive(true);
        
        log.info("Created knowledge category: {}", category.getId());
        return category;
    }

    public List<KnowledgeCategory> getCategories() {
        // Implementation for getting categories
        return Collections.emptyList();
    }

    public KnowledgeTag createTag(String name, String description) {
        KnowledgeTag tag = new KnowledgeTag();
        tag.setId(UUID.randomUUID().toString());
        tag.setName(name);
        tag.setDescription(description);
        tag.setCreatedAt(LocalDateTime.now());
        tag.setUsageCount(0);
        
        log.info("Created knowledge tag: {}", tag.getId());
        return tag;
    }

    public List<KnowledgeTag> getPopularTags(int limit) {
        // Implementation for getting popular tags
        return Collections.emptyList();
    }

    public KnowledgeAnalytics getAnalytics(String articleId) {
        KnowledgeAnalytics analytics = new KnowledgeAnalytics();
        analytics.setArticleId(articleId);
        analytics.setViewCount(0);
        analytics.setBookmarkCount(0);
        analytics.setAverageRating(0.0);
        analytics.setRatingCount(0);
        analytics.setSearchRank(0);
        analytics.setPopularityScore(0.0);
        
        return analytics;
    }

    public List<KnowledgeArticle> getTrendingArticles(int limit) {
        // Implementation for getting trending articles
        return Collections.emptyList();
    }

    public Map<String, Object> getSearchSuggestions(String query) {
        Map<String, Object> suggestions = new HashMap<>();
        suggestions.put("categories", Arrays.asList("Emergency Response", "Medical", "Logistics"));
        suggestions.put("tags", Arrays.asList("disaster", "emergency", "response", "training"));
        suggestions.put("articles", Arrays.asList("Basic First Aid", "Emergency Communication", "Resource Management"));
        
        return suggestions;
    }

    public KnowledgeArticle translateArticle(String articleId, String targetLanguage) {
        // Implementation for translating article
        KnowledgeArticle article = getArticle(articleId);
        article.setLanguage(targetLanguage);
        article.setTitle(article.getTitle() + " (Translated)");
        
        log.info("Translated article {} to {}", articleId, targetLanguage);
        return article;
    }

    public void deleteArticle(String articleId) {
        log.info("Deleted knowledge article: {}", articleId);
    }

    // Data classes
    public static class KnowledgeArticle {
        private String id;
        private String title;
        private String content;
        private String category;
        private List<String> tags;
        private String authorId;
        private String language;
        private boolean isPublic;
        private LocalDateTime createdAt;
        private LocalDateTime updatedAt;
        private int viewCount;
        private double rating;
        private ArticleStatus status;

        // Getters and setters
        public String getId() { return id; }
        public void setId(String id) { this.id = id; }

        public String getTitle() { return title; }
        public void setTitle(String title) { this.title = title; }

        public String getContent() { return content; }
        public void setContent(String content) { this.content = content; }

        public String getCategory() { return category; }
        public void setCategory(String category) { this.category = category; }

        public List<String> getTags() { return tags; }
        public void setTags(List<String> tags) { this.tags = tags; }

        public String getAuthorId() { return authorId; }
        public void setAuthorId(String authorId) { this.authorId = authorId; }

        public String getLanguage() { return language; }
        public void setLanguage(String language) { this.language = language; }

        public boolean isPublic() { return isPublic; }
        public void setPublic(boolean aPublic) { isPublic = aPublic; }
        // Compatibility setter for Lombok-style naming
        public void setIsPublic(boolean aPublic) { this.isPublic = aPublic; }

        public LocalDateTime getCreatedAt() { return createdAt; }
        public void setCreatedAt(LocalDateTime createdAt) { this.createdAt = createdAt; }

        public LocalDateTime getUpdatedAt() { return updatedAt; }
        public void setUpdatedAt(LocalDateTime updatedAt) { this.updatedAt = updatedAt; }

        public int getViewCount() { return viewCount; }
        public void setViewCount(int viewCount) { this.viewCount = viewCount; }

        public double getRating() { return rating; }
        public void setRating(double rating) { this.rating = rating; }

        public ArticleStatus getStatus() { return status; }
        public void setStatus(ArticleStatus status) { this.status = status; }
    }

    public static class KnowledgeCategory {
        private String id;
        private String name;
        private String description;
        private String parentCategoryId;
        private LocalDateTime createdAt;
        private boolean isActive;

        // Getters and setters
        public String getId() { return id; }
        public void setId(String id) { this.id = id; }

        public String getName() { return name; }
        public void setName(String name) { this.name = name; }

        public String getDescription() { return description; }
        public void setDescription(String description) { this.description = description; }

        public String getParentCategoryId() { return parentCategoryId; }
        public void setParentCategoryId(String parentCategoryId) { this.parentCategoryId = parentCategoryId; }

        public LocalDateTime getCreatedAt() { return createdAt; }
        public void setCreatedAt(LocalDateTime createdAt) { this.createdAt = createdAt; }

        public boolean isActive() { return isActive; }
        public void setActive(boolean active) { isActive = active; }
        // Compatibility setter for Lombok-style naming
        public void setIsActive(boolean active) { this.isActive = active; }
    }

    public static class KnowledgeTag {
        private String id;
        private String name;
        private String description;
        private LocalDateTime createdAt;
        private int usageCount;

        // Getters and setters
        public String getId() { return id; }
        public void setId(String id) { this.id = id; }

        public String getName() { return name; }
        public void setName(String name) { this.name = name; }

        public String getDescription() { return description; }
        public void setDescription(String description) { this.description = description; }

        public LocalDateTime getCreatedAt() { return createdAt; }
        public void setCreatedAt(LocalDateTime createdAt) { this.createdAt = createdAt; }

        public int getUsageCount() { return usageCount; }
        public void setUsageCount(int usageCount) { this.usageCount = usageCount; }
    }

    public static class ArticleRating {
        private String id;
        private String articleId;
        private String userId;
        private double rating;
        private String comment;
        private LocalDateTime createdAt;

        // Getters and setters
        public String getId() { return id; }
        public void setId(String id) { this.id = id; }

        public String getArticleId() { return articleId; }
        public void setArticleId(String articleId) { this.articleId = articleId; }

        public String getUserId() { return userId; }
        public void setUserId(String userId) { this.userId = userId; }

        public double getRating() { return rating; }
        public void setRating(double rating) { this.rating = rating; }

        public String getComment() { return comment; }
        public void setComment(String comment) { this.comment = comment; }

        public LocalDateTime getCreatedAt() { return createdAt; }
        public void setCreatedAt(LocalDateTime createdAt) { this.createdAt = createdAt; }
    }

    public static class ArticleBookmark {
        private String id;
        private String articleId;
        private String userId;
        private LocalDateTime createdAt;

        // Getters and setters
        public String getId() { return id; }
        public void setId(String id) { this.id = id; }

        public String getArticleId() { return articleId; }
        public void setArticleId(String articleId) { this.articleId = articleId; }

        public String getUserId() { return userId; }
        public void setUserId(String userId) { this.userId = userId; }

        public LocalDateTime getCreatedAt() { return createdAt; }
        public void setCreatedAt(LocalDateTime createdAt) { this.createdAt = createdAt; }
    }

    public static class KnowledgeAnalytics {
        private String articleId;
        private int viewCount;
        private int bookmarkCount;
        private double averageRating;
        private int ratingCount;
        private int searchRank;
        private double popularityScore;

        // Getters and setters
        public String getArticleId() { return articleId; }
        public void setArticleId(String articleId) { this.articleId = articleId; }

        public int getViewCount() { return viewCount; }
        public void setViewCount(int viewCount) { this.viewCount = viewCount; }

        public int getBookmarkCount() { return bookmarkCount; }
        public void setBookmarkCount(int bookmarkCount) { this.bookmarkCount = bookmarkCount; }

        public double getAverageRating() { return averageRating; }
        public void setAverageRating(double averageRating) { this.averageRating = averageRating; }

        public int getRatingCount() { return ratingCount; }
        public void setRatingCount(int ratingCount) { this.ratingCount = ratingCount; }

        public int getSearchRank() { return searchRank; }
        public void setSearchRank(int searchRank) { this.searchRank = searchRank; }

        public double getPopularityScore() { return popularityScore; }
        public void setPopularityScore(double popularityScore) { this.popularityScore = popularityScore; }
    }

    public enum ArticleStatus {
        DRAFT, PUBLISHED, ARCHIVED, DELETED
    }
}

File: frontend/src/services/knowledgeBaseService.ts
import { apiService } from './api';

export interface KnowledgeArticle {
  id: string;
  title: string;
  content: string;
  category: string;
  tags: string[];
  authorId: string;
  language: string;
  isPublic: boolean;
  createdAt: string;
  updatedAt: string;
  viewCount: number;
  rating: number;
  status: 'DRAFT' | 'PUBLISHED' | 'ARCHIVED' | 'DELETED';
}

export interface KnowledgeCategory {
  id: string;
  name: string;
  description: string;
  parentCategoryId?: string;
  createdAt: string;
  isActive: boolean;
}

export interface KnowledgeTag {
  id: string;
  name: string;
  description: string;
  createdAt: string;
  usageCount: number;
}

export interface KnowledgeAnalytics {
  articleId: string;
  viewCount: number;
  bookmarkCount: number;
  averageRating: number;
  ratingCount: number;
  searchRank: number;
  popularityScore: number;
}

export interface CreateArticleRequest {
  title: string;
  content: string;
  category: string;
  tags: string;
  language: string;
  isPublic: boolean;
}

export interface UpdateArticleRequest {
  title?: string;
  content?: string;
  category?: string;
  tags?: string;
  language?: string;
  isPublic?: boolean;
}

export interface RateArticleRequest {
  rating: number;
  comment?: string;
}

export interface CreateCategoryRequest {
  name: string;
  description: string;
  parentCategoryId?: string;
}

export interface CreateTagRequest {
  name: string;
  description: string;
}

class KnowledgeBaseService {
  private baseUrl = '/knowledge-base';

  async createArticle(request: CreateArticleRequest): Promise<KnowledgeArticle> {
    return apiService.post(`${this.baseUrl}/articles`, request);
  }

  async updateArticle(articleId: string, request: UpdateArticleRequest): Promise<KnowledgeArticle> {
    return apiService.put(`${this.baseUrl}/articles/${articleId}`, request);
  }

  async getArticle(articleId: string): Promise<KnowledgeArticle> {
    return apiService.get(`${this.baseUrl}/articles/${articleId}`);
  }

  async searchArticles(filters: {
    query?: string;
    category?: string;
    tags?: string[];
    language?: string;
    isPublic?: boolean;
    limit?: number;
  } = {}): Promise<KnowledgeArticle[]> {
    return apiService.get(`${this.baseUrl}/articles`, filters);
  }

  async getArticlesByCategory(category: string, limit: number = 50): Promise<KnowledgeArticle[]> {
    return apiService.get(`${this.baseUrl}/articles/category/${category}`, { limit });
  }

  async getPopularArticles(limit: number = 10): Promise<KnowledgeArticle[]> {
    return apiService.get(`${this.baseUrl}/articles/popular`, { limit });
  }

  async getRecentArticles(limit: number = 10): Promise<KnowledgeArticle[]> {
    return apiService.get(`${this.baseUrl}/articles/recent`, { limit });
  }

  async rateArticle(articleId: string, request: RateArticleRequest): Promise<KnowledgeArticle> {
    return apiService.post(`${this.baseUrl}/articles/${articleId}/rate`, request);
  }

  async bookmarkArticle(articleId: string): Promise<KnowledgeArticle> {
    return apiService.post(`${this.baseUrl}/articles/${articleId}/bookmark`);
  }

  async removeBookmark(articleId: string): Promise<void> {
    return apiService.delete(`${this.baseUrl}/articles/${articleId}/bookmark`);
  }

  async getUserBookmarks(): Promise<KnowledgeArticle[]> {
    return apiService.get(`${this.baseUrl}/bookmarks`);
  }

  async getRelatedArticles(articleId: string, limit: number = 5): Promise<KnowledgeArticle[]> {
    return apiService.get(`${this.baseUrl}/articles/${articleId}/related`, { limit });
  }

  async createCategory(request: CreateCategoryRequest): Promise<KnowledgeCategory> {
    return apiService.post(`${this.baseUrl}/categories`, request);
  }

  async getCategories(): Promise<KnowledgeCategory[]> {
    return apiService.get(`${this.baseUrl}/categories`);
  }

  async createTag(request: CreateTagRequest): Promise<KnowledgeTag> {
    return apiService.post(`${this.baseUrl}/tags`, request);
  }

  async getPopularTags(limit: number = 20): Promise<KnowledgeTag[]> {
    return apiService.get(`${this.baseUrl}/tags/popular`, { limit });
  }

  async getArticleAnalytics(articleId: string): Promise<KnowledgeAnalytics> {
    return apiService.get(`${this.baseUrl}/articles/${articleId}/analytics`);
  }

  async getTrendingArticles(limit: number = 10): Promise<KnowledgeArticle[]> {
    return apiService.get(`${this.baseUrl}/articles/trending`, { limit });
  }

  async getSearchSuggestions(query: string): Promise<Record<string, any>> {
    return apiService.get(`${this.baseUrl}/search/suggestions`, { query });
  }

  async translateArticle(articleId: string, targetLanguage: string): Promise<KnowledgeArticle> {
    return apiService.post(`${this.baseUrl}/articles/${articleId}/translate`, { targetLanguage });
  }

  async deleteArticle(articleId: string): Promise<void> {
    return apiService.delete(`${this.baseUrl}/articles/${articleId}`);
  }
}

export const knowledgeBaseService = new KnowledgeBaseService();


